server/src/main/java/org/apache/iotdb/db/query/fill/PreviousFill.java

```diff
@@ -61,26 +68,19 @@ public long getBeforeRange() {
 
   @Override
   public TimeValuePair getFillResult() throws IOException {
-    TimeValuePair beforePair = null;
-    TimeValuePair cachedPair;
-    while (batchData.hasCurrent() || allDataReader.hasNextBatch()) {
-      if (!batchData.hasCurrent() && allDataReader.hasNextBatch()) {
-        batchData = allDataReader.nextBatch();
-      }
-      cachedPair = new TimeValuePair(batchData.currentTime(), batchData.currentTsPrimitiveType());
-      batchData.next();
-      if (cachedPair.getTimestamp() <= queryTime) {
-        beforePair = cachedPair;
-      } else {
-        break;
-      }
-    }
+    return executor.getLastPoint();
+  }
 
-    if (beforePair != null) {
-      beforePair.setTimestamp(queryTime);
-    } else {
-      beforePair = new TimeValuePair(queryTime, null);
-    }
-    return beforePair;
+  @Override
+  public void constructReaders(Path path, QueryContext context)
+      throws StorageEngineException {
+    Filter timeFilter = constructFilter();
+    QueryDataSource queryDataSource =
+        QueryResourceManager.getInstance().getQueryDataSource(path, context, timeFilter);
+    // update filter by TTL
+    timeFilter = queryDataSource.updateFilterUsingTTL(timeFilter);
+    executor = new PreviousFillExecutor(
+        path, dataType, context, queryDataSource, timeFilter, null, null, queryTime);
```

Comment:
Move these to the constructor of PreviousFill, leave empty for the constructReaders. The LinearFill should also be refactored, then this method will be removed

Reply:
Removed, move these init code into configureFill